src/packages/hex.gleam

```diff
@@ -3,41 +3,204 @@ import gleam/json
+  // 3. Save new release in DB
+  // 4. Update package in DB if out of date
+  // 5. Return the new list of DB Package Type
+  try latest_release =
+    package.releases
+    |> list.at(0)
+    |> result.map_error(ItemNotInListError)
+
   let req =
     http.default_req()
     |> http.set_method(http.Get)
+    |> http.prepend_req_header("User-Agent", "GleamPackages")
     |> http.set_host("hex.pm")
-    |> http.set_path("/api/packages?sort=updated_at&page=1")
+    |> http.set_path(
+      "/api/packages/"
+      |> string.append(package.name)
+      |> string.append("/releases/")
+      |> string.append(latest_release.version),
+    )
 
   try response =
     hackney.send(req)
     |> result.map_error(HackneyError)
 
+  try release =
+    json.decode(response.body, full_release_decoder())
+    |> result.map_error(JsonError)
+
+  // Keep this package because the current version is made with gleam! ✨
+  case list.contains(release.meta.build_tools, "gleam") {
+    True -> Ok(package)
+    False -> Error(ItemNotInListError(Nil))
+  }
+}
+
+fn query_all_packages(
+  last_page: List(Package),
+  next_page: Int,
+  last_ran: Int,
+) -> Result(List(Package), Error) {
+  let req =
+    http.default_req()
+    |> http.set_method(http.Get)
+    |> http.prepend_req_header("User-Agent", "GleamPackages")
+    |> http.set_host("hex.pm")
+    |> http.set_path(
+      "/api/packages?sort=updated_at&page="
+      |> string.append(int.to_string(next_page)),
+    )
+
+  try response =
+    hackney.send(req)
+    |> result.map_error(HackneyError)
+
+  // The packages just fetched from the page
   try packages =
     json.decode(response.body, d.list(package_decoder()))
     |> result.map_error(JsonError)
 
-  io.debug(packages)
+  // The packages that have been updated since we last scanned and indexed the
+  // hex api.
+  let new_packages =
+    packages
+    |> list.filter(fn(x) {
+      parse_iso8601_to_gregorian_seconds(x.updated_at) > last_ran
+    })
 
-  Ok(1)
+  let all_packages = list.append(last_page, new_packages)
+
+  // Checks if the new_packages are the same length (or greater - even though 
+  // imposible in theory) than the packages fetched from hex. If it is that
+  // means that more packages need to be fetched as the entire page is new
+  // from the last scan, if not then we know we have got all the new packages
+  // and anything elase can be ignored as it should in theory have already been
+  // indexed.
+  case list.length(new_packages) >= list.length(packages) {
+    True -> query_all_packages(all_packages, next_page + 1, last_ran)
+    False -> Ok(all_packages)
+  }
 }
 
-type Package {
+pub type Package {
   Package(name: String, updated_at: String, releases: List(Release))
 }
 
-type Release {
+pub type Release {
   Release(version: String, url: String)
 }
 
+pub type ReleaseMeta {
+  ReleaseMeta(app: Option(String), build_tools: List(String))
+}
+
+pub type FullRelease {
+  FullRelease(version: String, url: String, meta: ReleaseMeta)
+}
```

Comment:
What's a full release vs a regular release?

Reply:
Models are going to be reworked when I start working with the DB Fully, names should make more sense after that.

Reply:
Refactored all the existsing types and have left room for DB types to be implemented, the names are now more explained and have comments to make it even easier 😄